fix(compiler-cli): downlevel angular decorators to static properties

devversion opened the pull request:
fix(compiler-cli): downlevel angular decorators to static properties

In v7 of Angular we removed `tsickle` from the default `ngc` pipeline.
This had the negative potential of breaking ES2015 output and SSR due
to a limitation in TypeScript.

TypeScript by default preserves type information for decorated constructor
parameters when `emitDecoratorMetadata` is enabled. For example,
consider this snippet below:

```ts
@Directive()
export class MyDirective {
  constructor(button: MyButton) {}
}

export class MyButton {}
```

TypeScript would generate metadata for the `MyDirective` class it has
a decorator applied. This metadata would be needed in JIT mode, or
for libraries that provide `MyDirective` through NPM. The metadata would
look as followed:

```js
let MyDirective = class MyDir {}

MyDirective = __decorate([
  Directive(),
  __metadata("design:paramtypes", [MyButton]),
], MyDirective);

let MyButton = class MyButton {}
```

Notice that TypeScript generated calls to `__decorate` and
`__metadata`. These calls are needed so that the Angular compiler
is able to determine whether `MyDirective` is actually an directive,
and what types are needed for dependency injection.

The limitation surfaces in this concrete example because `MyButton`
is declared after the `__metadata(..)` call, while `__metadata`
actually directly references `MyButton`. This is illegal though because
`MyButton` has not been declared at this point. This is due to the
so-called temporal dead zone in JavaScript. Errors like followed will
be reported at runtime when such file/code evaluates:

```js
Uncaught ReferenceError: Cannot access 'MyButton' before initialization
```

As noted, this is a TypeScript limitation because ideally TypeScript
shouldn't evaluate `__metadata`/reference `MyButton` immediately.
Instead, it should defer the reference until `MyButton` is actually
declared. This limitation will not be fixed by the TypeScript team
though because it's a limitation as per current design and they will
only revisit this once the tc39 decorator proposal is finalized
(currently stage-2 at time of writing).

Given this wontfix on the TypeScript side, and our heavy reliance on
this metadata in libraries (and for JIT mode), we intend to fix this
from within the Angular compiler by downleveling decorators to static
properties that don't need to evaluate directly. For example:

```js
MyDirective.ctorParameters = () => [MyButton];
```

With this snippet above, `MyButton` is not referenced directly. Only
lazily when the Angular runtime needs it. This mitigates the temporal
dead zone issue caused by a limitation in TypeScript's decorator
metadata output. See: https://github.com/microsoft/TypeScript/issues/27519.

In the past (as noted; before version 7), the Angular compiler by
default used tsickle that already performed this transformation. We
moved the transformation to the CLI for JIT and `ng-packager`, but now
we realize that we can move this all to a single place in the compiler
so that standalone ngc consumers can benefit too, and that we can
disable tsickle in our Bazel `ngc-wrapped` pipeline (that currently
still relies on tsickle to perform this decorator processing).

This transformation also has another positive side-effect of making
Angular application/library code more compatible with server-side
rendering. In principle, TypeScript would also preserve type information
for decorated class members (similar to how it did that for constructor
parameters) at runtime. This becomes an issue when your application
relies on native DOM globals for decorated class member types. e.g.

```ts
@Input() panelElement: HTMLElement;
```

Your application code would then reference `HTMLElement` directly
whenever the source file is loaded in NodeJS for SSR. `HTMLElement`
does not exist on the server though, so that will become an invalid
reference. One could work around this by providing global mocks for
these DOM symbols, but that doesn't match up with other places where
dependency injection is used for mocking DOM/browser specific symbols.

More context in this issue: #30586. The TL;DR here is that the Angular
compiler does not care about types for these class members, so it won't
ever reference `HTMLElement` at runtime.

Fixes #30106. Fixes #30586. Fixes #30141.
Resolves FW-2196. Resolves FW-2199.

devversion commented on goldens/size-tracking/aio-payloads.json:
```diff
@@ -21,8 +21,8 @@
-        "main-es2015": 429710,
-        "polyfills-es2015": 52195
+        "main-es2015": 431366,
+        "polyfills-es2015": 52655
```
* `main-es2015` previously was at `429841`, so this is an increase of `1.525kb`.
* `polyfills-es2015` remained as is. No change. Just updated it to reflect `master`.

I need to look into it and do some comparisons.

devversion replied:
Resolved by not using a TS trick to retain ctor parameter imports as done in CLI. The trick has the negative side-effect of preserving full imports (with potentially unused specifiers).

devversion commented:
@alxhub @IgorMinar I marked this as `master-only` because I think it's not critical to go into V10 given that we have re-enabled tsickle in `ng_package`. We surely could get it into v10 as a fix if desired 😄

alan-agius4 commented:
I think it’s best that this goes in v10, because otherwise we’ll be unable to remove this logic from ng-packagr and CLI until v11, that’s because ng-packagr and CLI are typically compatible for a particular major version of the framework.

Note: the transformer needs to be added as part of the public api, because in the CLI this transformer is only used when using JIT mode, which in CLI happens via TSC.

devversion commented:
@alan-agius4 Yeah, we chatted about that being exposed. I wonder how we can do that as compiler code never is public API as per `PUBLIC_API.md`. How has it been done in the past for similar things?

alan-agius4 commented:
Since the compiler code is not bundled we can opt to use deep imports in this case.

alan-agius4 reviewed (approved):
LGTM.

IgorMinar reviewed (changes requested):
Overall this looks really really good @devversion!

I left a few comments and I'm going to rely on @alxhub's review of the transformer code itself(I skipped that part during this review pass).

The biggest sticking point for me is `emitDecoratorMetadata`- see my longer comment about that.

With regards to your comment about "master-only" vs "master&patch". I think that we should get this in v10 to enable simplification of the downstream tooling as soon as possible. So please relabel the PR.

Lastly, and I'm very hesitant to even suggest it this late in the RC, but I do have to bring it up because otherwise we'll have to go through more churn and migrations: how about we rename `.decorators` and `ctorParams` to `ɵdec` and `ɵctor` (or something similar prefixed with `ɵ`). The goal of this would be to ensure that these props are clearly marked as private APIs and that we follow the same conventions for these apis as we do elsewhere.

The property name change needs more discussion, so please do not act on it, but think about it and we can discuss it during the FW sync meeting. 

(one known downstream change that would need to happen as a result of this renaming would be to update the CLI/build-optimizer passes to look for both the old and new property names, @alan-agius4 could weigh in on how difficult that would be and if there is more to it than an extra property lookup in AST)

alan-agius4 commented:
Renaming the props is not a big change from the CLI POV, I think the framework will be impacted more, because the compiler & ngcc will need to be aware of these new annotations for decorators.

Currently the compiler/cli/ngcc can process/reflect decorators that are either emitted either by tsickle or TypeScript.

devversion commented on packages/compiler-cli/integrationtest/tsconfig-build.json:
```diff
@@ -12,6 +12,7 @@
     "baseUrl": ".",
     "declaration": true,
     "experimentalDecorators": true,
+    "emitDecoratorMetadata": true,
```
Let's discuss this below in the overall `emitDecoratorMetadata` comment.

devversion commented on packages/compiler-cli/test/ngc_spec.ts:
```diff
@@ -654,9 +646,42 @@ describe('ngc transformer command-line', () => {
       });
+
+      it('should add metadata as static fields with type information if ' +
+             '`emitDecoratorMetadata` is enabled',
```
That is a good question. I've been thinking about this the last days a lot, and there are various reasons why this ended up like that:

TypeScript has a limitation in transformers that doesn't allow us to preserve imports without the `emitDecoratorMetadata` option. We need to preserve some imports that previously were type-only, but now are used as values in `ctorParameters`. TypeScript does refresh it's type checking state after transformers, so TypeScript will start eliding these imports incorrectly. e.g.

```ts
import {MatButton} from '@angular/material/button';

@Directive()
export class Test {
  constructor(button: MatButton) {}
}
```

In the example above: `MatButton` is type-only as per TypeScript's analysis. That is correct.. but not when we run our transformer that accesses `MatButton` as a value in `Test.ctorParameters = <..>`.

There are tricks that we can use to ensure that these imports are not elided. Though these have downsides. e.g. it could cause other unused imports to be preserved in the output too. In fact I had such trick applied in the initial version of this PR, but these retained imports can cause significant size increase as applications are not by default marked as `sideEffects: false`, and Webpack would bring in unnecessary chunks that are unused when build optimizer removes `ctorParameters`.

The CLI runs into a similar limitation when it removes decorators. It has a custom transformer for cleaning up unused imports later.. though it seems contradictory to implement such one in ngc too as it would basically mean that we re-implement TypeScript's logic for determining/eliding unused imports. We could certainly do that in NGC though.

An alternative workaround I came up with (one that will not prevent TS from eliding unused imports). This one would be the best if it wouldn't rely on some TS internals. It basically tells TypeScript that the given type symbols are now referenced as values, and their imports need to be retained. This matches the exact logic `emitDecoratorMetadata` has in TypeScript. That is actually why conceptually `emitDecoratorMetadata` needs to be set for type information.

This is aligned with how TypeScript emits decorators/metadata. e.g. by default, decorators are just downleveled, but if `emitDecoratorMetadata` is set, we also include type information.

The thinking here is that applications by default run in AOT mode and don't have `emitDecoratorMetadata` set (as done in CLI's).  In those cases, they wouldn't actually need the type information, nor the downleveled decorators. The CLI has a transformer for deleting decorators. Libraries on the other side want to include type information and would set `emitDecoratorMetadata`. That makes their library compatible for consumption in AOT and JIT.

Angular w/ tsickle worked around this in the past by just always setting `emitDecoratorMetadata = true` if decorators needed to be downleveled. These are the options from what I can tell:

1. Use the TS trick that keeps imports, but prevents eliding of other unused ones. We would then need another transform for deleting unused imports.. Basically re-implementing TypeScript's logic..
  The transformer would then most likely diverge a lot from the original one in tsickle.
2. Use the workaround that tells TypeScript about imported symbols that shouldn't be elided. No need to re-implement logic for deleting unused imports. Could be prone to internal changes in TypeScript as it relies on non-public API (TS doesn't export all the good things needed for good AST transformation!)
3. Always set `emitDecoratorMetadata = true` in NGC. This would match with what [we did before v7](https://github.com/angular/angular/blob/6.0.0/packages/compiler-cli/src/main.ts#L52) when tsickle still ran by default. I'm not sure if that has any impacts these days.
4. Require libraries or JIT consumers to explicitly set `emitDecoratorMetadata` when they want type information for DI included. That matches with how TypeScript emits metadata/decorator. 

I initially used (2) and it worked like a charm, but it was rather more like an experiment to see if we can do better than what has been done in the past of Angular with tsickle. Generally I'd definitely also prefer that we always include type information as that would be the most predictable way.

alxhub replied:
I think `ngc` (View Engine) should continue to assume `emitDecoratorMetadata` under the same circumstances that it did before. If `annotationsAs !== 'decorators'`, we turn on both our transform and `emitDecoratorMetadata`. I believe the absence of this logic is what's breaking g3 presubmit, since ngc_wrapped sets `annotationsAs = 'static fields'`.

clydin replied:
> Yeah, that sounds like option (3). Definitely not an ideal solution, but one with the fewest tricks required. One issue here would be that projects end up with type metadata for custom decorators, even though they had emitDecoratorMetadata = false. The CLI and build optimizer will only strip Angular decorators.

I think this is could turn out to be particularly problematic for libraries with custom decorators.  By leaving the metadata, the underlying TS issue could be triggered causing these libraries to no longer function.  I think `emitDecoratorMetadata` should be avoided completely if possible since its design is fundamentally flawed in ES2015+ code.

devversion replied:
@clydin Agreed. That's what I was referring too. Please feel free to also weigh in on the various other options being discussed here. I think your feedback would be appreciated!

alxhub replied:
I think the right approach is a modification of option 3:

`emitDecoratorMetadata` is automatically set by `ngc` unless the user has explicitly set it to `false`.

If `emitDecoratorMetadata` is set to `false` explicitly, no transformation happens - we emit the decorators as TS would have done.

If `emitDecoratorMetadata` is set (or coerced) to `true`, we add our transform and generate the full suite of metadata that `tsickle` used to.

As far as Bazel goes, its behavior should then be no different than plain `ngc`. We set `annotateForClosureCompiler` in `ngc_wrapped`, which causes `ngc` to emit via `tsickle`. Decorators are transformed internally (not in `tsickle`) because `emitDecoratorMetadata` will be assumed as above.

Does that sound reasonable? This way if applications/consumers decide they don't want decorator metadata, they can disable it explicitly.

clydin replied:
> If emitDecoratorMetadata is set to false explicitly, no transformation happens - we emit the decorators as TS would have done.

Wouldn't that break JIT mode for libraries that cannot enable the option?

devversion replied:
I had a long chat with Alex yesterday about this. We acknowledged that (3) and (4) are not good options as they would break libraries that use custom decorators. e.g.

```ts
@CustomDecorator()
export class SomeClass {
  constructor(button: Button) {}
}

export class Button {}
```

This currently could be shipped as part of libraries with ng-packagr, but when we'd require `ng-packagr` to enable `emitDecoratorMetadata`, it will break because the constructor type would be captured in the vanilla TS metadata and the [TypeScript temporal dead zone](https://github.com/microsoft/TypeScript/issues/27519) would surface.

Options remaining are (1) and (2). We are going with a mix of (1) and (2) where we properly tell TypeScript to preserve imports, but do not need a custom post-transform (as in the CLI) that cleans up unused imports. We can fully rely on TypeScript cleaning up unused/type-only imports. And the best thing is that we do not even need to rely on an internal TS API (though on some internal understanding of how the compiler works)

I will push these changes today. We should be able to get another review on this then.

alxhub commented on packages/bazel/src/ngc-wrapped/index.ts:
```diff
@@ -277,12 +259,23 @@ export function compile({
+  // By default, annotations for closure compiler are disabled in prodmode,
+  // unless the workspace is known to be `google3`.
+  if (!bazelOpts.es5Mode && compilerOpts.annotateForClosureCompiler === undefined) {
+    compilerOpts.annotateForClosureCompiler = bazelOpts.workspaceName === 'google3';
```
I prefer the former `if` statement that explicitly set `= true` or `= false` depending on the workspace name. This is trickier to follow.

IgorMinar replied:
funny. I actually thought the same. Then I tried to rewrite the code and ended up with just suggesting that the comment above is clarified. I'm fine with making this an explicit `true` / `false` assignment to improve the readability of the code.

devversion replied:
I don't have anything opposed to changing it back to the nested `if`. Seems more verbose for something super simple, but I'd agree that it's slightly more readable.

alxhub commented on packages/compiler-cli/src/transformers/program.ts:
```diff
@@ -99,8 +93,7 @@ class AngularCompilerProgram implements Program {
   private _optionsDiagnostics: Diagnostic[] = [];
-  // TODO(issue/24571): remove '!'.
-  private _reifiedDecorators!: Set<StaticSymbol>;
+  private _transformTsDiagnostics: ts.Diagnostic[] = [];
```
This is cleanly done, but sadly I think it's futile. This array only gets populated during `emit`, long after any diagnostics have been retrieved.

If you look at how this works with diagnostics coming from tsickle, the compiler drops them on the floor (it pays no attention to the return value of `emitWithTsickle`). We can just do the same.

devversion replied:
That's a fair point. I think we could incorporate these diagnostics into the emit result diagnostics, but I'm not sure if we even would want to report diagnostics (given that tsickle dropped them on the floor too).

If we'd now start printing diagnostics that could signify a breaking change. Changing it to ignore these diagnostics. Let me know if you think we should handle them as part of the emit result.

devversion committed:
fix(compiler-cli): downlevel angular decorators to static properties

In v7 of Angular we removed `tsickle` from the default `ngc` pipeline.
This had the negative potential of breaking ES2015 output and SSR due
to a limitation in TypeScript.

TypeScript by default preserves type information for decorated constructor
parameters when `emitDecoratorMetadata` is enabled. For example,
consider this snippet below:

```
@Directive()
export class MyDirective {
  constructor(button: MyButton) {}
}

export class MyButton {}
```

TypeScript would generate metadata for the `MyDirective` class it has
a decorator applied. This metadata would be needed in JIT mode, or
for libraries that provide `MyDirective` through NPM. The metadata would
look as followed:

```
let MyDirective = class MyDir {}

MyDirective = __decorate([
  Directive(),
  __metadata("design:paramtypes", [MyButton]),
], MyDirective);

let MyButton = class MyButton {}
```

Notice that TypeScript generated calls to `__decorate` and
`__metadata`. These calls are needed so that the Angular compiler
is able to determine whether `MyDirective` is actually an directive,
and what types are needed for dependency injection.

The limitation surfaces in this concrete example because `MyButton`
is declared after the `__metadata(..)` call, while `__metadata`
actually directly references `MyButton`. This is illegal though because
`MyButton` has not been declared at this point. This is due to the
so-called temporal dead zone in JavaScript. Errors like followed will
be reported at runtime when such file/code evaluates:

```
Uncaught ReferenceError: Cannot access 'MyButton' before initialization
```

As noted, this is a TypeScript limitation because ideally TypeScript
shouldn't evaluate `__metadata`/reference `MyButton` immediately.
Instead, it should defer the reference until `MyButton` is actually
declared. This limitation will not be fixed by the TypeScript team
though because it's a limitation as per current design and they will
only revisit this once the tc39 decorator proposal is finalized
(currently stage-2 at time of writing).

Given this wontfix on the TypeScript side, and our heavy reliance on
this metadata in libraries (and for JIT mode), we intend to fix this
from within the Angular compiler by downleveling decorators to static
properties that don't need to evaluate directly. For example:

```
MyDirective.ctorParameters = () => [MyButton];
```

With this snippet above, `MyButton` is not referenced directly. Only
lazily when the Angular runtime needs it. This mitigates the temporal
dead zone issue caused by a limitation in TypeScript's decorator
metadata output. See: https://github.com/microsoft/TypeScript/issues/27519.

In the past (as noted; before version 7), the Angular compiler by
default used tsickle that already performed this transformation. We
moved the transformation to the CLI for JIT and `ng-packager`, but now
we realize that we can move this all to a single place in the compiler
so that standalone ngc consumers can benefit too, and that we can
disable tsickle in our Bazel `ngc-wrapped` pipeline (that currently
still relies on tsickle to perform this decorator processing).

This transformation also has another positive side-effect of making
Angular application/library code more compatible with server-side
rendering. In principle, TypeScript would also preserve type information
for decorated class members (similar to how it did that for constructor
parameters) at runtime. This becomes an issue when your application
relies on native DOM globals for decorated class member types. e.g.

```
@Input() panelElement: HTMLElement;
```

Your application code would then reference `HTMLElement` directly
whenever the source file is loaded in NodeJS for SSR. `HTMLElement`
does not exist on the server though, so that will become an invalid
reference. One could work around this by providing global mocks for
these DOM symbols, but that doesn't match up with other places where
dependency injection is used for mocking DOM/browser specific symbols.

More context in this issue: #30586. The TL;DR here is that the Angular
compiler does not care about types for these class members, so it won't
ever reference `HTMLElement` at runtime.

Fixes #30106. Fixes #30586. Fixes #30141.
Resolves FW-2196. Resolves FW-2199.

clydin commented on packages/compiler-cli/src/transformers/downlevel_decorators_transform.ts:
```diff
@@ -0,0 +1,660 @@
+      // that TypeScript does not elide imports for such type references. Read more
+      // about this in the description for `patchAliasReferenceResolution`.
+      if (patchAliasReferenceResolution(
+              context as TransformationContextWithResolver, referencedParameterTypes) === false) {
```
I think `referencedParameterTypes` should be cleared per source file invocation.
The transformer can technically be reused to process more than one file.  TS doesn't appear to currently when using a `ts.Program` but it is setup to do so.
https://github.com/microsoft/TypeScript/blob/093e08396d6175520aad303dbacc32d7970e9fcd/src/compiler/transformer.ts#L137-L147

devversion replied:
The problem I'd see here though is that it could happen that we clear the list before TypeScript even ran its import eliding. We could solve that though by just re-declaring a new set. I assume that's what you meant?

devversion replied:
I think we should keep it as is because otherwise we'll have N sets whereas `N` is the amount of source files. In the worst case, a separate emit resolver will have a list of alias import declarations that it doesn't even check reference for (unless separate emit resolvers; but same AST nodes with parse tree; that would be seem contradictory though). What do you think?

clydin commented on packages/compiler-cli/src/transformers/downlevel_decorators_transform.ts:
```diff
@@ -0,0 +1,660 @@
+ * @param isClosureCompilerEnabled Whether closure annotations need to be added where needed.
+ */
+export function getDownlevelDecoratorsTransform(
+    typeChecker: ts.TypeChecker, diagnostics: ts.Diagnostic[], host: ReflectionHost,
```
Use of `ReflectionHost` may be problematic for CLI integration.  This transformer will only be used in JIT there and AOT compiler functionality is not present in that mode.

devversion replied:
The reflection host should just rely on the type-checker and we just use it for determining the origin of decorators (so we don't need to re-implement import resolution again). I don't think that should be an issue.

devversion replied:
@clydin Given that the reflection host should be fine with any arbitrary standard TS program, this should be good. We'll have a better factory entry-point for the CLI in `compiler-cli/tooling` where it only takes a TypeScript program and nothing more. Does that sound reasonable to you?

The plan would be then that this transform is used in the CLI, so that we have a canonical place for decorator/metadata downleveling. I'm not sure though how exactly it will work in the CLI. Could the CLI start depending on the compiler-cli exposed transformer. What happens if one uses an older compiler-cli version, but the latest CLI?

clydin replied:
Both the program instance and the type checker instance would be available when creating the transformer.
The FW and CLI are version locked by the major version.  So in this case CLI v10 can safely use anything that will be in FW v10.  However, if something was only available in FW v10.1 then the CLI could use it but would need a fallback if FW v10.0 was used.

devversion replied:
Got it. Sounds like we would be in good time to use this in the CLI. Thanks for the details!

alxhub commented on packages/compiler-cli/src/transformers/downlevel_decorators_transform.ts:
```diff
@@ -0,0 +1,660 @@
+
+/** Describes a subset of the TypeScript internal emit resolver. */
+type EmitResolver = {
+  isReferencedAliasDeclaration(node: ts.Node, checkChildren?: boolean): boolean;
```
This should be an optional method as well - we should be able to detect if it's not present.

devversion replied:
We're detecting it already, but just not having it typed here. I'll make it optional.

alxhub commented on packages/compiler-cli/src/transformers/downlevel_decorators_transform.ts:
```diff
@@ -0,0 +1,660 @@
+ *
+ * @returns A boolean indicating whether the alias reference solution could be patched properly.
+ */
+function patchAliasReferenceResolution(
```
1) Move this function into a separate file, so it's not lost in the rest of the transform code.
2) Instead of returning `boolean`, let's have it throw the error itself. We can rename it to `patchAliasReferenceResolutionOrDie`.
3) The error should be user-friendly and should explain that the version of TS is incompatible with the Angular compiler and they should downgrade (presumably, they're seeing it for the first time after updating TS).

alxhub commented on packages/compiler-cli/src/transformers/downlevel_decorators_transform.ts:
```diff
@@ -0,0 +1,660 @@
+ * solves it conceptually the same way, but obviously doesn't need to access an `@internal` API.
+ *
+ * See:
+ * https://sourcegraph.com/github.com/microsoft/TypeScript@3eaa7c65f6f076a08a5f7f1946fd0df7c7430259/-/blob/src/compiler/checker.ts#L31219-31257
```
Link directly to Github and not to sourcegraph.

devversion replied:
haha! good one 😛 I would have done that if the file wasn't ~2.4mb and could be displayed in Github. I figured adding a comment on why Github vs. Sourcegraph isn't necessary.

alxhub commented on packages/compiler-cli/src/transformers/downlevel_decorators_transform.ts:
```diff
@@ -0,0 +1,660 @@
+ * @returns A boolean indicating whether the alias reference solution could be patched properly.
+ */
+function patchAliasReferenceResolution(
+    context: TransformationContextWithResolver, referencedAliases: Set<ts.Declaration>): boolean {
```
The public API of this function should accept a `ts.TransformationContext` and cast internally.

clydin reviewed (changes requested):
Nice job integrating this.

Couple additional comments:
* I think the property name changes are a good idea but agree with @IgorMinar that they can wait.
* Along with the renaming, I think that static methods should be used instead of static properties where appropriate (e.g., `ctorParameters` which is really a static method in disguise).  All static properties need to be down-leveled so their use is giving more work to the compiler during each build/rebuild.
* How will the integration into the CLI for JIT support look?

devversion commented on packages/compiler-cli/src/transformers/downlevel_decorators_transform.ts:
```diff
@@ -0,0 +1,660 @@
+     * downleveled static properties. Also collects constructor type metadata for
+     * class declaration that are decorated with an Angular decorator.
+     */
+    function decoratorDownlevelVisitor(node: ts.Node): ts.Node {
+      switch (node.kind) {
+        case ts.SyntaxKind.ClassDeclaration: {
+          return transformClassDeclaration(node as ts.ClassDeclaration);
+        }
+        default:
+          return ts.visitEachChild(node, decoratorDownlevelVisitor, context);
+      }
+    }
```
Sure. I'll go through an clean up all the original tsickle code, but note that as per filecomment, this is the _original_ tsickle code with small modifications (no clean ups)

devversion commented on packages/compiler-cli/src/transformers/downlevel_decorators_transform.ts:
```diff
@@ -0,0 +1,660 @@
+      // Downlevel decorators and constructor parameter types. We will keep track of all
+      // referenced constructor parameter types so that we can instruct TypeScript to
+      // not elide their imports if they previously were only type-only.
+      return decoratorDownlevelVisitor(sf) as ts.SourceFile;
```
Same as below. I just ported this from the tsickle transform w/ small modifications (no cleanup). I'll clean this up though.

devversion commented on packages/compiler-cli/src/transformers/downlevel_decorators_transform.ts:
```diff
@@ -0,0 +1,660 @@
+      }
+      if (!toLower.length) return [undefined, element, []];
+
+      if (!element.name || element.name.kind !== ts.SyntaxKind.Identifier) {
```
Same as below. I intentionally didn't clean things up, as the goal was to keep modifications minimal in the tsickle port. I'll clean this up based on discussion with Alex though.

clydin replied:
Understood.  There's probably a decent about of additional refactoring that would be good long-term especially with newer TS capabilities (most of this was written against TS 2.x).
But I noticed a could small changes that could remove casting (which can be very error prone) so figured I'd point them out.  Feel free to ignore them for now if preferred.

devversion replied:
Yeah, that makes sense. Definitely appreciated. Thanks! I figured I'd just make it clear that initial goal was to not clean this up, but I agree that there are a lot of possibilities 😄

devversion commented on packages/compiler-cli/src/transformers/downlevel_decorators_transform.ts:
```diff
@@ -0,0 +1,660 @@
+ * Either it's used in @angular/core, or it's imported from there.
+ */
+function isAngularDecorator(decorator: Decorator, isCore: boolean): boolean {
+  return isCore || (decorator.import !== null && decorator.import.from === '@angular/core');
```
Yeah, I think this is a general issue also in ngtsc. We should definitely consider this there.

devversion commented on packages/compiler-cli/src/transformers/downlevel_decorators_transform.ts:
```diff
@@ -0,0 +1,660 @@
+  if (originalReferenceResolution === undefined) {
+    return false;
+  }
+  emitResolver.isReferencedAliasDeclaration = function(node, checkChildren) {
+    if (isAliasImportDeclaration(node) && referencedAliases.has(node)) {
+      return true;
+    }
+    return originalReferenceResolution.call(emitResolver, node, checkChildren);
+  };
```
I like it. 👍  thanks!

devversion commented on packages/compiler-cli/src/transformers/downlevel_decorators_transform.ts:
```diff
@@ -0,0 +1,660 @@
+  // If the `getEmitResolver` method is not available, TS most likely changed the
+  // internal structure of the transformation context. We will abort gracefully.
+  if (context.getEmitResolver === undefined) {
+    return false;
```
FWIW: We are throwing below. This has been moved into the method now for clarity though

devversion commented on packages/compiler-cli/src/transformers/downlevel_decorators_transform.ts:
```diff
@@ -0,0 +1,660 @@
+ */
+function isAliasImportDeclaration(node: ts.Node): node is ts.ImportSpecifier|ts.NamespaceImport|
+    ts.ImportClause {
+  return ts.isImportSpecifier(node) || ts.isNamespaceImport(node) || ts.isImportClause(node);
```
Yeah, that is correct. Let's keep them in the list of referenced alias declarations for safety and completeness though. Shouldn't hurt.

clydin commented on packages/compiler-cli/src/transformers/program.ts:
```diff
@@ -591,18 +519,26 @@ class AngularCompilerProgram implements Program {
       flatModuleMetadataTransforms.push(transformer);
     }
 
-    if (stripDecorators) {
```
Is Ivy not stripping decorators anymore with this change?

devversion replied:
This flag was only `true` for `program.emitRender3`. Though that is no longer the real entry-point for Ivy as it uses its own program. I removed it while being at it.

devversion committed:
fixup! fix(compiler-cli): downlevel angular decorators to static properties

Address feedback

clydin commented on packages/compiler-cli/src/transformers/patch_alias_reference_resolution.ts:
```diff
@@ -0,0 +1,119 @@
+};
+
+/** Describes a subset of the TypeScript internal emit resolver. */
+type EmitResolver = {isReferencedAliasDeclaration ? (node: ts.Node, checkChildren?: boolean) :
```
nit: maybe an interface for this to improve the formatting.

alxhub replied:
I agree with this, for both types.

alxhub commented on packages/compiler-cli/src/transformers/program.ts:
```diff
@@ -7,26 +7,28 @@
-import {createBundleIndexHost, MetadataCollector, ModuleMetadata} from '../metadata';
+import {translateDiagnostics} from '../diagnostics/translate_diagnostics';
+import {createBundleIndexHost, MetadataCollector} from '../metadata';
+import {isAngularCorePackage} from '../ngtsc/core/src/compiler';
```
ngtsc tries to ensure that all imports go through the barrel `index.ts` files - so this should be:
```suggestion
import {isAngularCorePackage} from '../ngtsc/core';
```
You might need to add a re-export to make that work.

devversion committed:
fixup! fix(compiler-cli): downlevel angular decorators to static properties

Use barrel imports

alxhub commented:
[Presubmit](http://test/OCL:314993213:BASE:314993260:1591391321435:7367c7d1)

IgorMinar reviewed (approved):
LGTM now. Thank you! Excellent work @devversion! ❤️

IgorMinar commented:
[presubmit](http://test/OCL:314731511:BASE:315028346:1591404948218:a3455f1c)

IgorMinar reviewed (approved):
LGTM! Thank you!

Reviewed-for: global-approvers

devversion committed:
fixup! fix(compiler-cli): downlevel angular decorators to static properties

Fix incorrect import reference with conflicting local parameter name

devversion commented on packages/bazel/test/ng_package/example_package.golden:
```diff
@@ -611,7 +611,7 @@ let MyService = /** @class */ (() => {
 })();
 export { MyService };
-//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoicHVibGljLWFwaS5qcyIsInNvdXJjZVJvb3QiOiIiLCJzb3VyY2VzIjpbIi4uLy4uLy4uLy4uLy4uLy4uLy4uLy4uLy4uL3BhY2thZ2VzL2JhemVsL3Rlc3QvbmdfcGFja2FnZS9leGFtcGxlL2ltcG9ydHMvcHVibGljLWFwaS50cyJdLCJuYW1lcyI6W10sIm1hcHBpbmdzIjoiQUFBQTs7Ozs7O0dBTUc7QUFFSCxPQUFPLEVBQUMsVUFBVSxFQUFDLE1BQU0sZUFBZSxDQUFDO0FBQ3pDLE9BQU8sRUFBQyxlQUFlLEVBQUMsTUFBT ...
+//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoicHVibGljLWFwaS5qcyIsInNvdXJjZVJvb3QiOiIiLCJzb3VyY2VzIjpbIi4uLy4uLy4uLy4uLy4uLy4uLy4uLy4uLy4uL3BhY2thZ2VzL2JhemVsL3Rlc3QvbmdfcGFja2FnZS9leGFtcGxlL2ltcG9ydHMvcHVibGljLWFwaS50cyJdLCJuYW1lcyI6W10sIm1hcHBpbmdzIjoiQUFBQTs7Ozs7O0dBTUc7QUFFSCxPQUFPLEVBQUMsVUFBVSxFQUFDLE1BQU0sZUFBZSxDQUFDO0FBQ3pDLE9BQU8sRUFBQyxlQUFlLEVBQUMsTUFBT ...
```
Note for reviewer: You might wonder why this changed the source map. This is because before this fix, the actual parameter type identifier was used for `ctorParameters`. Now since we use the import alias though for `ctorParameters`, the source map will change back to what it was with tsickle too.

alxhub reviewed (approved):
LGTM with the fix for aliased constructor parameter names.

alxhub commented:
[Presubmit](http://test/OCL:315509583:BASE:315509629:1591723487449:ab74d3d)

devversion committed:
fixup! fix(compiler-cli): downlevel angular decorators to static properties

Properly handle namespace imports resolving to non-value

alxhub commented:
[And again](http://test/OCL:315535390:BASE:315535448:1591730591781:f0118361)

devversion committed:
fixup! fix(compiler-cli): downlevel angular decorators to static properties

alxhub commented:
[Here we go again](http://test/OCL:315568666:BASE:315568709:1591740426265:5efd5946)
